src/include/plan_node/abstract_plan_node.h

```diff
@@ -0,0 +1,141 @@
+    return children_[child_index].get();
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Accessors
+  //===--------------------------------------------------------------------===//
+
+  // Each sub-class will have to implement this function to return their type
+  // This is better than having to store redundant types in all the objects
+  virtual PlanNodeType GetPlanNodeType() const { return PlanNodeType::ABSTRACTPLAN; }
+
+  // Get a string representation for debugging
+  virtual const std::string GetInfo() const;
+
+  // Get the output schema for the plan node. The output schema contains information on columns of the output of
+  // the plan node operator
+  std::shared_ptr<OutputSchema> GetOutputSchema() const { return output_schema_; }
+
+  // Get the estimated cardinality of this plan
+  int GetEstimatedCardinality() const { return estimated_cardinality_; }
+
+  // FOR TESTING ONLY. This function should only be called during construction of plan (ConvertOpExpression) or
+  // for tests.
+  void SetEstimatedCardinality(int cardinality) { estimated_cardinality_ = cardinality; }
+
+  //===--------------------------------------------------------------------===//
+  // JSON Serialization/Deserialization
+  //===--------------------------------------------------------------------===//
+
+  /**
+   * Return the current plan node in JSON format.
+   * @return JSON representation of plan node
+   */
+  virtual nlohmann::json ToJson() const;
+
+  /**
+   * Populates the plan node with the information in the given JSON.
+   * Undefined behavior occurs if the JSON has a different PlanNodeType.
+   */
+  virtual void FromJson(const nlohmann::json &json);
+
+  //===--------------------------------------------------------------------===//
+  // Utilities
+  //===--------------------------------------------------------------------===//
+  virtual common::hash_t Hash() const {
+    common::hash_t hash = GetOutputSchema()->Hash();
+    hash = common::HashUtil::CombineHashes(hash, common::HashUtil::Hash(GetPlanNodeType()));
+    for (auto &child : GetChildren()) {
+      hash = common::HashUtil::CombineHashes(hash, child->Hash());
+    }
+    return hash;
+  }
+
+  virtual bool operator==(const AbstractPlanNode &rhs) const {
+    if (*GetOutputSchema() != *rhs.GetOutputSchema()) return false;
+    auto num = GetChildren().size();
+    if (num != rhs.GetChildren().size()) return false;
+    for (unsigned int i = 0; i < num; i++) {
+      if (*GetChild(i) != *const_cast<AbstractPlanNode *>(rhs.GetChild(i))) return false;
+    }
+    return true;
+  }
+
+  virtual bool operator!=(const AbstractPlanNode &rhs) const { return !(*this == rhs); }
+
+  virtual std::unique_ptr<AbstractPlanNode> Copy() const = 0;
+
+ private:
+  // A plan node can have multiple children
+  std::vector<std::unique_ptr<AbstractPlanNode>> children_;
+
+  int estimated_cardinality_ = 500000;
```

Comment:
do we really want this default value here?

Reply:
This was a hack in original peloton, I removed it once and it caused the bloomfilter to crash. It should probably be safe to remove in terrier and given that we're going to have a new execution engine.